Make it easier to run CI jobs locally

Kobzol opened the pull request:
This PR extends the Python CI script to perform a poor man's CI-like execution of a given CI job locally. It's not perfect, but it's better than nothing.

r? @jieyouxu

Kobzol committed:
Rename CI script from `calculate-job-matrix` to `ci.py`

Kobzol committed:
Add a command to run a given Linux CI job locally

Kobzol committed:
Update documentation

lqd commented:
(if this lands, https://rustc-dev-guide.rust-lang.org/tests/docker.html#testing-with-docker will also need to be updated)

Kobzol committed:
Rename `image` property of CI jobs to `name`

The `image` part didn't really make sense, especially since we started splitting CI jobs.

Kobzol commented on src/ci/docker/README.md:
````diff
@@ -1,29 +1,20 @@
-```
-
-for example:
+To run a specific CI job locally, you can use the following script:
 
 ```
-./src/ci/docker/run.sh x86_64-gnu
````
Good point, because "job" and "image" here isn't the same thing :)

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12538990642/job/34964689383) [(plain)](https://github.com/rust-lang/rust/commit/bdbf7830651ac92400d042106662c8f7e34d41ff/checks/34964689383/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

jieyouxu commented on src/ci/github-actions/ci.py:
```diff
@@ -27,13 +27,18 @@
     """
-    Add a `name` attribute to each job, based on its image and the given `prefix`.
+    Modify the `name` attribute of each job, based on its base name and the given `prefix`.
+    Add an `image` attribute to each job, base don its image.
```
Nit: `don` -> `on`

jieyouxu reviewed (approved):
It seems to work for me locally, nice :+1:

jieyouxu commented:
The changes LGTM, but I'm not an infra reviewer, so r? infra

Kobzol commented on src/ci/github-actions/ci.py:
```diff
@@ -181,30 +183,76 @@ def format_run_type(run_type: WorkflowRunType) -> str:
+    jobs = [job for job in jobs if job.get("image") == job_name]
+    if len(jobs) == 0:
+        raise Exception(f"Job `{job_name}` not found")
+    job = jobs[0]
```
It can pick a job with the same name if a PR and auto job have the same name, which can happen. I ordered the jobs so that auto jobs have priority; I didn't want to complicate the CLI by allowing users to select between PR and auto jobs. But it's true that they might not actually have the same config even if they have the same name, so maybe we should let users choose.

Kobzol committed:
Make it possible to select PR vs auto jobs

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12557821503/job/35011044198) [(plain)](https://github.com/rust-lang/rust/commit/c017f3f8903c346d2b3f907eac9f8436cb7e00a5/checks/35011044198/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.906 Building wheels for collected packages: reuse
#13 2.907   Building wheel for reuse (pyproject.toml): started
#13 3.152   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.153   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.153   Stored in directory: /tmp/pip-ephem-wheel-cache-pn_mroi3/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.156 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.547 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.548 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.083 Collecting virtualenv
#13 4.083 Collecting virtualenv
#13 4.132   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.290      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 27.4 MB/s eta 0:00:00
#13 4.349 Collecting platformdirs<5,>=3.9.1
#13 4.356   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.396 Collecting filelock<4,>=3.12.2
#13 4.424 Collecting distlib<1,>=0.3.7
#13 4.431   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.443      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 49.9 MB/s eta 0:00:00
#13 4.443      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 49.9 MB/s eta 0:00:00
#13 4.523 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.707 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      120768 kB
DirectMap2M:     6170624 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5752 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/src/ci/github-actions/ci.py:246: line longer than 100 chars
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Dec 31 09:29:10 UTC 2024
  network time: Tue, 31 Dec 2024 09:29:10 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

MarcoIeni commented on src/ci/github-actions/ci.py:
```diff
@@ -181,30 +188,96 @@ def format_run_type(run_type: WorkflowRunType) -> str:
+
+    custom_env = {}
+    custom_env["DEPLOY"] = "1"
+    custom_env.update({k: str(v) for (k, v) in job.get("env", {}).items()})
```
```suggestion
    # add to `custom_env` the environment variables of the job
    custom_env.update({k: str(v) for (k, v) in job.get("env", {}).items()})
```

Kobzol committed:
Apply review comments

Kobzol committed:
Update rustc-dev-guide

Kobzol commented:
Rebased, applied review comments and also updated rustc-dev-guide.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12656535291/job/35269333355) [(plain)](https://github.com/rust-lang/rust/commit/2f63bbe3a5928fd51d8124db2230f5f8d980ee34/checks/35269333355/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.722 Building wheels for collected packages: reuse
#13 2.723   Building wheel for reuse (pyproject.toml): started
#13 2.931   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.931   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132718 sha256=f1d7d99e2187cfcc743e73878997db8829d4c560719802a1230bd160f7d9cfca
#13 2.932   Stored in directory: /tmp/pip-ephem-wheel-cache-p68xfrud/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.934 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.332 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.332 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.864 Collecting virtualenv
#13 3.864 Collecting virtualenv
#13 3.902   Downloading virtualenv-20.28.1-py3-none-any.whl (4.3 MB)
#13 3.960      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 77.3 MB/s eta 0:00:00
#13 4.014 Collecting platformdirs<5,>=3.9.1
#13 4.017   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.035 Collecting distlib<1,>=0.3.7
#13 4.045      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 91.4 MB/s eta 0:00:00
#13 4.080 Collecting filelock<4,>=3.12.2
#13 4.083   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.083   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.163 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.344 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.1
#13 DONE 4.4s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      133056 kB
DirectMap2M:     7206912 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5761 files
tidy check
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking python file formatting
Would reformat: /checkout/src/ci/github-actions/ci.py
1 file would be reformatted, 27 files already formatted
1 file would be reformatted, 27 files already formatted

python formatting does not match! Printing diff:
error: unexpected argument '--check' found

  tip: to pass '--check' as a value, use '-- --check'

Usage: ruff check <FILES|--fix|--no-fix|--unsafe-fixes|--no-unsafe-fixes|--show-source|--no-show-source|--show-fixes|--no-show-fixes|--diff|--watch|--fix-only|--no-fix-only|--ignore-noqa|--output-format <OUTPUT_FORMAT>|--output-file <OUTPUT_FILE>|--target-version <TARGET_VERSION>|--preview|--no-preview|--select <RULE_CODE>|--ignore <RULE_CODE>|--extend-select <RULE_CODE>|--extend-ignore <RULE_CODE>|--per-file-ignores <PER_FILE_IGNORES>|--extend-per-file-ignores <EXTEND_PER_FILE_IGNORES>|--exclude <FILE_PATTERN>|--extend-exclude <FILE_PATTERN>|--fixable <RULE_CODE>|--unfixable <RULE_CODE>|--extend-fixable <RULE_CODE>|--extend-unfixable <RULE_CODE>|--respect-gitignore|--no-respect-gitignore|--force-exclude|--no-force-exclude|--line-length <LINE_LENGTH>|--dummy-variable-rgx <DUMMY_VARIABLE_RGX>|--no-cache|--cache-dir <CACHE_DIR>|--stdin-filename <STDIN_FILENAME>|--extension <EXTENSION>|--exit-zero|--exit-non-zero-on-fix|--statistics|--add-noqa|--show-files|--show-settings|--ecosystem-ci>
For more information, try '--help'.
For more information, try '--help'.
rerun tidy with `--extra-checks=py:fmt --bless` to reformat Python code
some tidy checks failed
tidy error: checks with external tool 'ruff' failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Jan  7 17:58:05 UTC 2025
  network time: Tue, 07 Jan 2025 17:58:05 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

MarcoIeni commented:
Sorry for the delay, I lost the GitHub notification somehow!

@bors r+

bors commented:
:pushpin: Commit 65819b193ae59cc1696eb796681df4f397900e67 has been approved by `MarcoIeni`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=MarcoIeni 65819b193ae59cc1696eb796681df4f397900e67 -->
<!-- homu: {"type":"Approved","sha":"65819b193ae59cc1696eb796681df4f397900e67","approver":"MarcoIeni","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Kobzol commented:
Note to myself: next time we should mark CI affecting PRs with rollup=never :)
